File: desktop/tests/reauth.test.ts
import { describe, it, expect } from 'vitest';
import { needsReauth } from '../src/lib/reauth';

describe('needsReauth', () => {
  it('returns false when disabled regardless of clock', () => {
    expect(needsReauth({ enabled: false, lastAuthAtMs: null, nowMs: 0, windowMs: 0 })).toBe(false);
    expect(needsReauth({ enabled: false, lastAuthAtMs: 0, nowMs: 9e9, windowMs: 1000 })).toBe(false);
  });

  it('returns true when never authed this session', () => {
    expect(needsReauth({ enabled: true, lastAuthAtMs: null, nowMs: 5000, windowMs: 1000 })).toBe(true);
  });

  it('returns false within the grace window', () => {
    expect(needsReauth({ enabled: true, lastAuthAtMs: 1000, nowMs: 1500, windowMs: 1000 })).toBe(false);
  });

  it('returns true at exactly the window boundary (>=)', () => {
    expect(needsReauth({ enabled: true, lastAuthAtMs: 1000, nowMs: 2000, windowMs: 1000 })).toBe(true);
  });

  it('returns true past the window', () => {
    expect(needsReauth({ enabled: true, lastAuthAtMs: 1000, nowMs: 5000, windowMs: 1000 })).toBe(true);
  });

  it('windowMs of 0 always prompts when enabled (every write)', () => {
    expect(needsReauth({ enabled: true, lastAuthAtMs: 1000, nowMs: 1000, windowMs: 0 })).toBe(true);
  });
});

File: android/vault-access/src/test/kotlin/org/secretary/browse/ReauthTest.kt
package org.secretary.browse

import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class ReauthTest {
    private val window = ReauthWindow.V1_DEFAULT_MS

    @Test
    fun `null lastAuth always needs reauth`() {
        assertTrue(needsReauth(lastAuthAtMs = null, nowMs = 0L, windowMs = window))
    }

    @Test
    fun `within the window does not need reauth`() {
        assertFalse(needsReauth(lastAuthAtMs = 1_000L, nowMs = 1_000L + 29_999L, windowMs = window))
    }

    @Test
    fun `exactly at the window boundary needs reauth (inclusive)`() {
        assertTrue(needsReauth(lastAuthAtMs = 1_000L, nowMs = 1_000L + window, windowMs = window))
    }

    @Test
    fun `past the window needs reauth`() {
        assertTrue(needsReauth(lastAuthAtMs = 1_000L, nowMs = 1_000L + window + 1L, windowMs = window))
    }

    @Test
    fun `a zero window always needs reauth (prompt before every write)`() {
        assertTrue(needsReauth(lastAuthAtMs = 5_000L, nowMs = 5_000L, windowMs = 0L))
    }

    @Test
    fun `the v1 default window is 30 seconds`() {
        org.junit.jupiter.api.Assertions.assertEquals(30_000L, ReauthWindow.V1_DEFAULT_MS)
    }
}
